### ISSUE TITLE
"key-spacing" align colon issue with comma-first

### ISSUE THREAD

**Issue by woodyrew** [eslint/eslint#3877]
ESLint version: **v1.5.0**

`.eslintrc`:
```json
{
  "rules": {
    "key-spacing": [2, {"align": "colon", "beforeColon": false, "afterColon": true}]
  }
}
```

Failing source code - `fail.js`:
```js
var obj = {
    foo : foo
  , bar : bar
  , cats: cats
};
```

Result:
```bash
$ eslint fail.js 

/path/to/fail.js
  2:5  error  Missing space after key "foo"  key-spacing

✖ 1 problem (1 error, 0 warnings)
```

The error disappears when an extra two spaces are added before the first colon in the object.
e.g.
`no_fail.js`:
```js
var obj = {
    foo   : foo
  , bar : bar
  , cats: cats
};
```

Am I missing something here? If more information is required, let me know.

Thanks.

**Comment by ilyavolodin** [comment]
It looks like it might be related to comma-first style. (2 extra spaces would match comma+space)

**Comment by woodyrew** [comment]
I agree, but the `key-spacing` rule shouldn't care where the comma is placed as long as the parameter name and colons line up.

**Comment by IanVS** [comment]
I think the problem is that we assume the lines are on the same level of indentation, which in your case they are not.  If your indentation is not consistent, it's very hard to do any vertical aligning.

**Comment by woodyrew** [comment]
That's part of the problem, the assumption that the indent level is where the parameter name starts but not limited to. It seems to assume the parameter name is the length from the first non-whitespace character.

Using the npm coding style: https://docs.npmjs.com/misc/coding-style#comma-first is even worse.
e.g.
```js
var obj = { foo : foo  // Extra space after key "foo".
          , bar : bar  // Missing space after key "bar".
          , cats: cats // Missing space after key "cats".
};
```
No errors:
```js
var obj = { foo: foo
          , bar          : bar
          , cats         : cats
};
```

**Comment by ilyavolodin** [comment]
@BYK @btmills any ideas if this is something we want to fix and how hard would it be?

**Comment by BYK** [comment]
My hunch is that this not being too hard to fix (and I think this is something we should fix). I'm way behind my e-mail and ESLint stuff though so can't promise a timely patch. Will see what I can do over the weekend (or close to it).

**Comment by woodyrew** [comment]
@ilyavolodin @BYK Thanks. If I can help in any way, let me know.

**Comment by BYK** [comment]
@woodyrew you can try submitting a PR as soon as this gets approved. Feel free to ping me or anyone else on the Gitter channel (use @-mentions to make sure you ping someone) for help if you ever need it :)

**Comment by nzakas** [comment]
Up to @btmills on whether to accept this

**Comment by btmills** [comment]
Very well, accepted. I'm not positive, but [`getFirstNodeInLine`](https://github.com/eslint/eslint/blob/master/lib/rules/key-spacing.js#L106) seems like a pretty good place to start looking for the culprit.

**Comment by btmills** [comment]
Working on this.

**Comment by shammellee** [comment]
It doesn't seem to be fixed. I know there's an extra space after the `a` key but that's what's required to align the colons with comma-first syntax.

**Code**
```js
let obj = {
  a  : 'foo',
  ,b : 'bar',
  ,c : 'baz',
};
```

**Error**
```
error   Extra space after key 'a'    key-spacing
```

**.eslintrc**
```
rules:
  comma-style:
    - error
    - first
  key-spacing:
    - error
    - singleLine: {}
      multiLine:
        beforeColon: true
        align: colon
```

**Comment by btmills** [comment]
@shammellee your `a` key being out of alignment with `b` and `c` is throwing this off. Your best bet would be to move the commas back so the keys can be aligned:

```js
let obj = {
  a : 'foo'
, b : 'bar'
, c : 'baz'
};
```

Or insert one extra space before the `a` to make room for the commas on subsequent lines:

```js
let obj = {
   a : 'foo'
  ,b : 'bar'
  ,c : 'baz'
};
```

If neither of those solutions is workable, can you open a new issue so we can discuss more there?
